File: src/main/scala/temportalist/compression/main/client/model/BakedCompressed.scala
package temportalist.compression.main.client.model

import java.util

import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.block.model.SimpleBakedModel.Builder
import net.minecraft.client.renderer.block.model.{BakedQuad, IBakedModel, ItemCameraTransforms, ItemOverrideList}
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.util.{BlockRenderLayer, EnumFacing}
import net.minecraft.util.math.BlockPos
import net.minecraftforge.client.MinecraftForgeClient
import net.minecraftforge.common.property.IExtendedBlockState
import temportalist.compression.main.common.Compression
import temportalist.compression.main.common.lib.{BlockProperties, EnumTier}

/**
  *
  * Created by TheTemportalist on 4/14/2016.
  *
  * @author TheTemportalist
  */
class BakedCompressed(private val overlays: Array[TextureAtlasSprite]) extends IBakedModel {

	private val overrideList = new ItemListCompressed(overlays)

	override def getQuads(state: IBlockState, side: EnumFacing, rand: Long): util.List[BakedQuad] = {
		state match {
			case extended: IExtendedBlockState =>
				val sampleStack = extended.getValue(BlockProperties.ITEMSTACK_UN)
				if (sampleStack == null) {
					//Compression.log("ERROR: " + state.toString + " has null inner stack!")
					return new util.ArrayList[BakedQuad]()
				}
				val sampleBlock = Block.getBlockFromItem(sampleStack.getItem)
				val sampleState = sampleBlock.getStateFromMeta(sampleStack.getItemDamage)
				val sampleModel = Minecraft.getMinecraft.getBlockRendererDispatcher.
						getBlockModelShapes.getModelForState(sampleState)

				return (MinecraftForgeClient.getRenderLayer match {
					case BlockRenderLayer.SOLID => sampleModel
					case BlockRenderLayer.TRANSLUCENT =>
						val i = EnumTier.getTierForSize(extended.getValue(BlockProperties.LONG_UN)).ordinal()
						val overlayModel = new Builder(sampleState, sampleModel, overlays(i), BlockPos.ORIGIN).makeBakedModel()
						overlayModel
					case _ => null
				}).getQuads(state, side, rand)

			case _ =>
		}
		new util.ArrayList[BakedQuad]()
	}

	override def isAmbientOcclusion: Boolean = true

	override def isBuiltInRenderer: Boolean = false

	override def isGui3d: Boolean = true

	override def getItemCameraTransforms: ItemCameraTransforms = ItemCameraTransforms.DEFAULT

	override def getParticleTexture: TextureAtlasSprite = overlays.last

	override def getOverrides: ItemOverrideList = this.overrideList

}

File: src/main/java/com/temportalist/compression/client/BakedCompressed.java
package com.temportalist.compression.client;

import com.google.common.collect.Lists;
import com.temportalist.compression.common.lib.BlockProperties;
import com.temportalist.compression.common.lib.EnumTier;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.model.BakedQuad;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ItemCameraTransforms;
import net.minecraft.client.renderer.block.model.ItemOverrideList;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.client.MinecraftForgeClient;
import net.minecraftforge.client.model.PerspectiveMapWrapper;
import net.minecraftforge.common.model.TRSRTransformation;
import net.minecraftforge.common.property.IExtendedBlockState;
import net.minecraft.client.renderer.block.model.SimpleBakedModel.Builder;
import org.apache.commons.lang3.tuple.Pair;

import javax.annotation.Nullable;
import javax.vecmath.Matrix4f;
import javax.vecmath.Vector3f;
import java.util.List;
import com.google.common.collect.ImmutableMap;

public class BakedCompressed implements IBakedModel {

    private TextureAtlasSprite[] overlays;
    private ItemOverrideList overrideList;

    public BakedCompressed(TextureAtlasSprite[] overlays) {
        this.overlays = overlays;
        this.overrideList = new ItemListCompressed(overlays);
    }

    @Override
    public ItemOverrideList getOverrides() {
        return this.overrideList;
    }

    @Override
    public List<BakedQuad> getQuads(@Nullable IBlockState state, @Nullable EnumFacing side, long rand) {
        if (state instanceof IExtendedBlockState) {
            IExtendedBlockState extended = (IExtendedBlockState)state;

            ItemStack sampleStack = extended.getValue(BlockProperties.ITEMSTACK_UN);
            if (sampleStack == null) {
                return Lists.newArrayList();
            }

            Block sampleBlock = Block.getBlockFromItem(sampleStack.getItem());
            IBlockState sampleState = sampleBlock.getStateFromMeta(sampleStack.getItemDamage());
            IBakedModel sampleModel = Minecraft.getMinecraft().getBlockRendererDispatcher().getBlockModelShapes().getModelForState(sampleState);

            IBakedModel layerModel;
            BlockRenderLayer layer = MinecraftForgeClient.getRenderLayer();
            if (layer == null) layerModel = sampleModel;
            else
            switch (layer) {
                case SOLID:
                    layerModel = sampleModel;
                    break;
                case TRANSLUCENT:
                    int i = EnumTier.getTier(extended.getValue(BlockProperties.TIER_UN)).ordinal();
                    layerModel = new Builder(sampleState, sampleModel, overlays[i], BlockPos.ORIGIN).makeBakedModel();
                    break;
                default:
                    layerModel = sampleModel;
                    break;
            }

            return layerModel.getQuads(state, side, rand);
        }
        return Lists.newArrayList();
    }

    @Override
    public boolean isAmbientOcclusion() { return true; }

    @Override
    public boolean isBuiltInRenderer() { return false; }

    @Override
    public boolean isGui3d() { return true; }

    @Override
    public TextureAtlasSprite getParticleTexture() {
        return this.overlays[this.overlays.length - 1];
    }

}
